src/fs/share.ts

```diff
@@ -0,0 +1,134 @@
+import * as cbor from "@ipld/dag-cbor"
+import { SymmAlg } from "keystore-idb/lib/types.js"
+
+import * as basic from "./protocol/basic.js"
+import * as crypto from "../crypto/index.js"
+import * as dataRoot from "../data-root.js"
+import * as ipfs from "../ipfs/basic.js"
+import * as pathing from "../path.js"
+import * as protocol from "./protocol/index.js"
+import * as shareKey from "./protocol/shared/key.js"
+
+import { Branch, DirectoryPath } from "../path.js"
+import { SharedBy, ShareDetails } from "./types.js"
+import { didToPublicKey } from "../did/transformers.js"
+import BareTree from "./bare/tree.js"
+import PrivateFile from "./v1/PrivateFile.js"
+import PrivateTree from "./v1/PrivateTree.js"
+import RootTree from "./root/tree.js"
+
+
+// CONSTANTS
+
+
+export const EXCHANGE_PATH: DirectoryPath = pathing.directory(
+  pathing.Branch.Public,
+  ".well-known",
+  "exchange"
+)
+
+
+
+// FUNCTIONS
+
+
+export async function privateNode(
+  rootTree: RootTree,
+  items: Array<[string, PrivateTree | PrivateFile]>,
+  { shareWith, sharedBy }: {
+    shareWith: string | string[],
+    sharedBy: SharedBy
+  }
+): Promise<ShareDetails> {
+  const exchangeDIDs = Array.isArray(shareWith)
+    ? shareWith
+    : shareWith.startsWith("did:")
+      ? [ shareWith ]
+      : await listExchangeDIDs(shareWith)
+
+  const counter = rootTree.sharedCounter || 1
+  const mmpt = rootTree.mmpt
+
+  // Create share keys
+  const shareKeysWithDIDs: [string, string][] = await Promise.all(exchangeDIDs.map(async did => {
+    return [
+      did,
+      await shareKey.create({
+        counter,
+        recipientExchangeDid: did,
+        senderRootDid: sharedBy.rootDid
+      })
+    ] as [string, string]
+  }))
+
+  // Create entry index
+  const indexKey = await crypto.aes.genKeyStr()
+  const index = await PrivateTree.create(mmpt, indexKey, null)
+
+  await Promise.all(
+    items.map(async ([name, item]) => {
+      const privateName = await item.getName()
+      return index.insertSoftLink({
+        key: item.key,
+        name,
+        privateName,
+        username: sharedBy.username
+      })
+    })
+  )
+
+  // Add entry index to ipfs
+  const symmKeyAlgo = SymmAlg.AES_GCM
+  const indexNode = Object.assign({}, index.header)
+  const indexResult = await basic.putFile(
+    await crypto.aes.encrypt(
+      new TextEncoder().encode( JSON.stringify(indexNode) ),
+      index.key,
+      symmKeyAlgo
+    )
+  )
+
+  // Create share payload
+  const payload = cbor.encode(shareKey.payload({
+    entryIndexCid: indexResult.cid,
```

Comment:
This entry index is not being pinned by the server because it can't reach it.
We need to find another place to put this CID so the server pins it.

Reply:
I thought so too but that doesn't seem like a good idea to put them there in their current form, since the `index` contains the AES keys with the pointers to that data (soft links). I guess one could then go through one's MMPT and extract that content?

Reply:
Oh no wait, I forgot the index was also encrypted with AES 🤦 
I'll put it in the MMPT 👍

Reply:
Added and now working fine across different accounts ✅

Reply:
@icidasset

Correct, it goes in the private partition 👍 Yes we _definitely_ need the index encrypted!

Also down the road it will make sense to put _unencrypted_ data in the same structure ("unlisted" as opposed to public/private), but that's down the road.

---

@matheus23 

Sure — I have no strong feelings one way or the other. You end up with a CID either way (namefilters point to CIDs), and you may get some speedup _when the link is missing_ by using namefilters... but that's not the typical case when you've had something shared with you. You can think of the shared link as being an assertion that the content is there, but the one layer of indirection can prevent you from spinning forever waiting to resolve a CID that doens't exist.